flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/DefaultHashFunctionProvider.java

```diff
@@ -0,0 +1,96 @@
+public class DefaultHashFunctionProvider implements HashFunctionProvider {
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public HashFunction getHashFunction(TableId tableId, Schema schema) {
+        return new DefaultHashFunction(schema);
+    }
+
+    static class DefaultHashFunction implements HashFunction {
+        private final List<RecordData.FieldGetter> primaryKeyGetters;
+
+        public DefaultHashFunction(Schema schema) {
+            primaryKeyGetters = createFieldGetters(schema);
+        }
+
+        @Override
+        public Integer apply(DataChangeEvent event) {
+            List<Object> objectsToHash = new ArrayList<>();
+            // Table ID
+            TableId tableId = event.tableId();
+            Optional.ofNullable(tableId.getNamespace()).ifPresent(objectsToHash::add);
+            Optional.ofNullable(tableId.getSchemaName()).ifPresent(objectsToHash::add);
+            objectsToHash.add(tableId.getTableName());
+
+            // Primary key
+            RecordData data =
+                    event.op().equals(OperationType.DELETE) ? event.before() : event.after();
+            for (RecordData.FieldGetter primaryKeyGetter : primaryKeyGetters) {
+                objectsToHash.add(primaryKeyGetter.getFieldOrNull(data));
+            }
+
+            // Calculate hash
+            return (Objects.hash(objectsToHash.toArray()) * 31) & 0x7FFFFFFF;
+        }
+
+        private List<RecordData.FieldGetter> createFieldGetters(Schema schema) {
+            List<RecordData.FieldGetter> fieldGetters =
+                    new ArrayList<>(schema.primaryKeys().size());
+            int[] primaryKeyPositions =
+                    schema.primaryKeys().stream()
+                            .mapToInt(
+                                    pk -> {
+                                        int i = 0;
+                                        while (!schema.getColumns().get(i).getName().equals(pk)) {
+                                            ++i;
+                                        }
+                                        if (i >= schema.getColumnCount()) {
+                                            throw new IllegalStateException(
+                                                    String.format(
+                                                            "Unable to find column \"%s\" which is defined as primary key",
+                                                            pk));
+                                        }
+                                        return i;
+                                    })
+                            .toArray();
+            for (int primaryKeyPosition : primaryKeyPositions) {
+                fieldGetters.add(
+                        RecordData.createFieldGetter(
+                                schema.getColumns().get(primaryKeyPosition).getType(),
+                                primaryKeyPosition));
+            }
```

Comment:
Chain this with previous `Stream<String>`?

Reply:
Sure.